Mathlib/Topology/MetricSpace/Lipschitz.lean

```diff
@@ -581,6 +590,127 @@ end Metric
+protected lemma comp  {f : β → γ} {g : α → β}
+    (hf : LocallyLipschitz f) (hg : LocallyLipschitz g) : LocallyLipschitz (f ∘ g) := by
+  intro x
+  -- g is Lipschitz on t ∋ x, f is Lipschitz on u ∋ g(x)
+  rcases hg x with ⟨Kg, t, ht, hgL⟩
+  rcases hf (g x) with ⟨Kf, u, hu, hfL⟩
+  -- idea: shrink t to ensure it is mapped to u
+  -- more precisely: restrict g to t' := t ∩ g⁻¹(u); the preimage of u under g':=g∣t.
+  let g' := t.restrict g
+  set t' : Set α := ↑(g' ⁻¹' u) with ht'
+  have h₁ : t' = t ∩ g ⁻¹' u := by
+    rw [ht']
+    ext1 y
+    simp [Lean.Internal.coeM]
+    aesop
+  have h₂ : t' ∈ 𝓝 x := by -- FIXME: surely, there is a nicer proof
+    -- by ht, t contains an open subset U
+    rcases (mem_nhds_iff.mp ht) with ⟨U, hUt, hUopen, hxU⟩
+    -- similarly, u contains an open subset V
+    rcases (mem_nhds_iff.mp hu) with ⟨V, hVt, hVopen, hgxV⟩
+    -- by continuity, g⁻¹(u) contains the open subset g⁻¹(V)
+    have : ContinuousOn g U := (hgL.mono hUt).continuousOn
+    have h : IsOpen (U ∩ (g ⁻¹' V)) := this.preimage_open_of_open hUopen hVopen
+    have : U ∩ (g ⁻¹' V) ⊆ t' := by rw [h₁]; apply inter_subset_inter hUt (preimage_mono hVt)
+    -- now, U ∩ g⁻¹(V) is an open subset contained in t'
+    rw [mem_nhds_iff]
+    use U ∩ (g ⁻¹' V)
+    exact ⟨this, ⟨h, ⟨hxU, hgxV⟩⟩⟩
+  have : g '' t' ⊆ u := by calc g '' t'
+    _ = g '' (t ∩ g ⁻¹' u) := by rw [h₁]
+    _ ⊆ g '' t ∩ g '' (g ⁻¹' u) := by apply image_inter_subset
+    _ ⊆ g '' t ∩ u := by gcongr; apply image_preimage_subset
+    _ ⊆ u := inter_subset_right _ _
+  use Kf * Kg, t'
+  exact ⟨h₂, hfL.comp (hgL.mono coe_subset) (mapsTo'.mpr this)⟩
+
+/-- If `f` and `g` are locally Lipschitz, so is the induced map `f × g` to the product type. -/
+protected lemma prod {f : α → β} (hf : LocallyLipschitz f) {g : α → γ} (hg : LocallyLipschitz g) :
+    LocallyLipschitz fun x => (f x, g x) := by
+  intro x
+  rcases hf x with ⟨Kf, t₁, h₁t, hfL⟩
+  rcases hg x with ⟨Kg, t₂, h₂t, hgL⟩
+  use max Kf Kg, t₁ ∩ t₂
+  constructor
+  · exact Filter.inter_mem h₁t h₂t
+  · intro y hy z hz
+    have h₁ : edist (f y) (f z) ≤ Kf * edist y z := hfL.mono (inter_subset_left t₁ t₂) hy hz
+    have h₂ : edist (g y) (g z) ≤ Kg * edist y z := hgL.mono (inter_subset_right t₁ t₂) hy hz
+    rw [ENNReal.coe_mono.map_max, Prod.edist_eq, ENNReal.max_mul]
+    exact max_le_max h₁ h₂
+
+protected theorem prod_mk_left (a : α) : LocallyLipschitz (Prod.mk a : β → α × β) :=
+  (LipschitzWith.prod_mk_left a).locallyLipschitz
+
+protected theorem prod_mk_right (b : β) : LocallyLipschitz (fun a : α => (a, b)) :=
+  (LipschitzWith.prod_mk_right b).locallyLipschitz
+
+protected theorem iterate {f : α → α} (hf : LocallyLipschitz f) : ∀ n, LocallyLipschitz f^[n]
+  | 0 => by simpa only [pow_zero] using LocallyLipschitz.id
+  | n + 1 => by rw [iterate_add, iterate_one]; exact (hf.iterate n).comp hf
+
+protected theorem mul {f g : Function.End α} (hf : LocallyLipschitz f)
```

Comment:
I would rather expect `LocallyLipschitz.mul` to mean the following: if `f` and `g` are two locally Lipschitz functions valued in the real line (or more generally any space with a multiplication and a distance that interact nicely), then their pointwise product is also locally Lipschitz. Same thing for `pow` below. Maybe rename your lemmas to `mul_end` and `pow_end` if you really need them?

Reply:
I have renamed the lemmas accordingly. I have also renamed the lemmas on LipschitzWith. (Having pow and mul mean one thing for Lipschitz functions and another for locally Lipschitz ones seems like the worst of all worlds.)

Reply:
Thinking again, I'm on the fence about this: on one hand, it makes perfect sense to reuse this lemma `LocallyLipschitz.mul_end` and to reserve `LocallyLipschitz.mul` for the result you describe.

On the other hand, the point-wise lemma `LocallyLipschitz.mul` is false for general Lipschitz functions (it requires boundedness): renaming would keep the name `LipschitzWith.mul` free, which we couldn't fill satisfactorily.

Reply:
To summarize, I see three options.
(A) Keep this lemma name. Counterargument: might be surprising; would need a rename when the `mul` result is added.
(B) Rename `LocallyLipschitz.mul`, but keep `LipschitzWith.mul`. Counter-argument. Is inconsistent.
(C) Rename also `LipschitzWith.mul`. This *is* consistent, but means `LipschitzWith.mul` cannot be filled without extra hypotheses. (If the rename passes the build, it shows that `LipschitzWith.mul` is unused - which might be evidence this is ok.)
(D) Delete this lemma. Not sure if this makes sense.

@sgouezel What do you think?

Reply:
I definitely want to keep `LocallyLipschitz.mul` for the fact that the pointwise product of locally Lipschitz functions is locally Lispchitz (this would match all the other lemmas `Continuous.mul`, `Differentiable.mul` and so on). Accordingly, I would like `LipschitzWith.mul` to say something like: if two functions are Lipschitz and bounded, then their pointwise product is Lipschitz.

So I see two reasonable options:
(1): use `LipschitzWith.mul_end` and `LocallyLipschitzWith.mul_end`.
(2): delete these lemmas.

(I don't know why these lemmas were introduced in the first place -- we don't have similar lemmas for other categories. Are they used anywhere in the library?)

Your choice between (1) and (2). I would do (1) because it is nicer to the person who introduced these lemmas, but do as you prefer.

Reply:
Makes sense. (I didn't check either why these lemmas were introduced.) I prefer (1) as well, and have thusly approved this PR.